container/state/state.go

```diff
@@ -266,16 +262,76 @@ func (s *State) SetRemovalInProgress() bool {
 	return false
 }
 
-// ResetRemovalInProgress make the RemovalInProgress state to false.
-func (s *State) ResetRemovalInProgress() {
+// ResetRemovalInProgressLocking make the RemovalInProgress state to false.
+func (s *State) ResetRemovalInProgressLocking() {
 	s.Lock()
 	s.RemovalInProgress = false
 	s.Unlock()
 }
 
-// SetDead sets the container state to "dead"
-func (s *State) SetDead() {
+// SetDeadLocking sets the container state to "dead"
+func (s *State) SetDeadLocking() {
+	s.Lock()
+	defer s.Unlock()
+	s.State = Dead
+	close(s.waitChan) // fire waiters for stop
+	s.waitChan = make(chan struct{})
+}
+
+// SetRawState allows setting any state user specified
+func (s *State) SetRawState(cs Cstate) {
+	s.State = cs
+	close(s.waitChan) // fire waiters for stop
+	s.waitChan = make(chan struct{})
+}
+
+// SetRawStateLocking allows setting any state user specified
+func (s *State) SetRawStateLocking(cs Cstate) {
 	s.Lock()
-	s.Dead = true
+	s.SetRawState(cs)
 	s.Unlock()
 }
+
+// IsRunning returns whether the running flag is set.
+// Used by Container to check whether a container is running
+func (s *State) IsRunning() bool {
+	return s.InState(Running)
+}
+
+// IsRunningLocking locks State and checks whether running flag is set
+func (s *State) IsRunningLocking() bool {
+	return s.InStateLocking(Running)
+}
+
+// IsPaused returns whether the paused flag is set.
+// Used by Container to check whether a container is paused
+func (s *State) IsPaused() bool {
+	return s.InState(Paused)
+}
+
+// IsPausedLocking locks State and checks whether paused flag is set
+func (s *State) IsPausedLocking() bool {
+	return s.InStateLocking(Paused)
+}
+
+// IsRestarting returns whether the restarting flag is set.
+// Used by Container to check whether a container is restarting
+func (s *State) IsRestarting() bool {
+	return s.InState(Restarting)
+}
+
+// IsRestartingLocking locks State and checks whether restarting flag is set
+func (s *State) IsRestartingLocking() bool {
+	return s.InStateLocking(Restarting)
+}
+
+// IsDead returns whether the dead flag is set.
+// Used by Container to check whether a container is dead
+func (s *State) IsDead() bool {
+	return s.InState(Dead)
+}
+
+// IsDeadLocking locks State and checks whether dead flag is set
+func (s *State) IsDeadLocking() bool {
```

Comment:
Yes, I also dislike it :-(. It sounds like some "dead lock" is gonna to happen.
I once considered to use "IsDeadL", "IsRunningL" but they are also horrible, hard to know its meaning from single function name. 
So I have to leave it here :(